experimental/CliffordAlgebras/src/CliffordAlgebras.jl

```diff
@@ -0,0 +1,584 @@
+
+Return the element `y` in the Clifford algebra containing $x$ such that $ay = x$,
+if it exists. Otherwise an error is raised.
+"""
+divexact(x::CliffordAlgebraElem, a::T) where {T<:RingElem} =
+  parent(x)(divexact.(coefficients(x), a))
+
+divexact(x::CliffordAlgebraElem, a::T) where {T<:Number} = parent(x)(divexact.(coefficients(x), a))
+
+################################################################################
+#
+#  Equality and hash
+#
+################################################################################
+
+function Base.:(==)(x::CliffordAlgebraElem{T}, y::CliffordAlgebraElem{T}) where {T}
+  @req parent(x) === parent(y) "The inputs must lie in the same Clifford algebra"
+  return coefficients(x) == coefficients(y)
+end
+
+function Base.hash(x::CliffordAlgebraElem, h::UInt)
+  b = 0x1c4629b4de23b24c % UInt
+  h = hash(parent(x), h)
+  h = hash(coefficients(x), h)
+  return xor(h, b)
+end
+
+################################################################################
+#
+#  Graded parts
+#
+################################################################################
+
+@doc raw"""
+    even_part(x::CliffordAlgebraElem) -> CliffordAlgebraElem
+
+Return the projection of $x$ onto the even Clifford algebra
+"""
+even_part(x::CliffordAlgebraElem) = parent(x)(even_coefficients(x))
+
+@doc raw"""
+    odd_part(x::CliffordAlgebraElem) -> CliffordAlgebraElem
+
+Return the projection of $x$ onto the odd Clifford algebra.
+"""
+odd_part(x::CliffordAlgebraElem) = parent(x)(odd_coefficients(x))
+
+################################################################################
+#
+#  Auxillary functions
+#
+################################################################################
+
+function _set_even_odd_coefficients!(x::CliffordAlgebraElem)
+  x.even_coeffs = map(
+    y -> if sum(digits(y - 1; base=2, pad=_dim_qf(x.parent))) % 2 == 0
+      x.coeffs[y]
+    else
+      x.parent.base_ring()
+    end, 1:(x.parent.dim)
+  )
+  x.odd_coeffs = x.coeffs - x.even_coeffs
+  return x
+end
+
+function __can_convert_coefficients(coeff::Vector{R}, K::Field) where {R}
+  if length(coeff) == 0
+    return true
+  end
+  try
+    K.(coeff)  # Try converting the coefficient vector to K
+    true
+  catch
+    false
+  end
+end
+
+_dim_qf(C::CliffordAlgebra) = ncols(C.gram)
+
+function _mul_aux(x::Vector{T}, y::Vector{T}, gram::MatElem{T}, i::Int) where {T}
+  if length(y) == 1 #|| is_zero(y)
+    return x .* y[1]
+  end
+  return _mul_aux(_mul_with_gen(x, i, gram), y[2:2:end], gram, i + 1) +
+         _mul_aux(x, y[1:2:end], gram, i + 1)
+end
+
+#Implements the right multiplication with 'i'-th generator e_i of the Clifford algebra containing x.
+#The result is returned as a coefficient vector for further computations.
+_mul_with_gen(x::Vector{T}, i::Int, gram::MatElem{T}) where {T} = sum(
+  map(
+    j ->
+      x[j] .*
+      _mul_baseelt_with_gen(BitVector(digits(j - 1; base=2, pad=ncols(gram))), i, gram),
```

Comment:
Converting `Int` values into `BitVector` seems very inefficient to me. Can't we do we without it?

Reply:
Thanks, of course we can. Changed in this [commit](https://github.com/oscar-system/Oscar.jl/pull/4370/commits/00b2cb1354677d596d83bfb70606be1b6eb3ce7b).